transport/src/main/java/io/netty/channel/AutoFlushHandler.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.channel;
+
+/**
+ * Netty is extremely flexible in providing a way to batch writes on the physical socket by providing a means to write
+ * and flush writes separately. This works for the usecases when the application pattern is one that provides
+ * predictable finite streams of data. In cases, when an infinite stream of data is to be written on a channel which
+ * does not have any predictable length and duration of writes, a user has the following choices:
+ * <ol>
+ * <li>Flush every write. </li>
+ * <li>Flush batches based on time and count (eg: Flush every 10 message or 1 second) . </li>
+ * </ol>
+ *
+ * Neither of the above is ideal as the nature of stream is completely unpredictable.
+ * Flushing every write is costly as flush is a system call.
+ * Flushing batches is arbitrary and has to be deviced for each stream in an arbitrary way.  
+ *
+ * This handler addresses the above usecase by providing a way to batch writes and flush them at the following points:
+ * <ol>
+ * <li>At the end of the eventloop iteration.</li>
+ * <li>If the channel becomes unwritable determined via {@link #channelWritabilityChanged(ChannelHandlerContext)}
+ * callback.</li>
+ * </ol>
+ *
+ * Any explicit flushes are honored as usual.
+ *
+ * <h2>Restrictions</h2>
+ *
+ * Due to limitation of existing APIs, {@link AutoFlushHandler} can only be used if the following conditions are met:
+ *
+ * <ol>
+ *     <li>Channel is registered with an {@link EventLoop} that extends {@link SingleThreadEventLoop}.</li>
+ *     <li>{@link ChannelPipeline} is an instance of {@link DefaultChannelPipeline}</li>
+ *     <li>This handler is not executed by any executor apart from the channel's eventloop.</li>
+ * </ol>
+ *
+ * If any of the above conditions are not met, this handler will fail on
+ * {@link #channelRegistered(ChannelHandlerContext)}.
+ *
+ * <h2>Recommendations</h2>
+ *
+ * The {@link AutoFlushHandler} should be put as first {@link ChannelHandler} in the
+ * {@link ChannelPipeline} to have the best effect.
+ */
+public class AutoFlushHandler extends ChannelDuplexHandler implements Runnable {
```

Comment:
@normanmaurer I should have mentioned this in this PR before, sorry about that.

So, here is the problem/reason why this handler can't be in the handler module:

- Logic to "wakeup on write without flush" exists in `DefaultChannelPipeline` and `AbstractChannelHandlerContext`.
- In order to determine whether auto-flush is enabled and hence wakeup should be issued can only be determined by checking if `AutoFlushHandler` is added to the pipeline.
- Since `DefaultChannelPipeline` exists in transport module which do not depend on handler module (correctly so), I can not refer to `AutoFlushHandler`.
- `AutoFlushHandler` needs to use `DefaultChannelPipeline.postAutoFlush()` method which is package-private.

So, one solution could be that `DefaultChannelPipeline` provide public methods `postAutoFlush()`, `autoFlushEnabled()` and `autoFlushDisable()` which can be invoked from `AutoFlushHandler` when required. I am not sure what you feel about making these methods public. If ok, I will go ahead and make that change.